test/make_dist.py

```diff
@@ -47,5 +50,104 @@ def make_dist(devel_mode):
     return subprocess.check_output(["make", "dump-dist"], universal_newlines=True).strip()
 
 
+def download_dist():
+    '''Download dists tarball for current git SHA from GitHub
+
+    These are produced by .github/workflows/build-dist.yml for every PR and push.
+    This is a lot faster than having to npm install and run webpack.
+
+    Returns path to downloaded tarball, or None if it isn't available.
+    This can happen because the current directory is not a git checkout, or it is
+    a SHA which is not pushed/PRed, or there is no ~/.config/github-token available.
+    '''
+    try:
+        sha = subprocess.check_output(["git", "rev-parse", "HEAD"], stderr=subprocess.DEVNULL).decode().strip()
+    except subprocess.CalledProcessError:
+        print("make_dist: not a git repository")
+        return None
+
+    dists = glob.glob(f"cockpit-*{sha[:8]}*.tar.xz")
+    if dists:
+        print("make_dist: already downloaded " + dists[0])
+        return os.path.abspath(dists[0])
+
+    sys.path.append(os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "bots"))
+    import task
+
+    # task.api defaults to the current checkout's origin, but the artifacts are on cockpit-project
+    # except if a developer wants to change the artifact building workflow on their fork, support that with $GITHUB_BASE
+    api = task.github.GitHub(repo=os.getenv("GITHUB_BASE", "cockpit-project/cockpit"))
+
+    # downloading GitHub artifacts requires a token
+    if not api.token:
+        print("make_dist: no GitHub API token available")
+        return None
+
+    # iterate artifacts and search our SHA
+    page = 1
+    batch_size = 100
+    download_url = None
+    while not download_url and batch_size == 100:
+        batch = api.get(f"actions/artifacts?per_page={batch_size}&page={page}")["artifacts"]
+        for artifact in batch:
+            if artifact["name"] == "dist-" + sha:
+                download_url = artifact["archive_download_url"]
+                break
+        # if the current batch is < 100, we have the last page and can stop
+        batch_size = len(batch)
+        page += 1
+
+    if not download_url:
+        print(f"make_dist: no download available for commit {sha}")
+        return None
+
+    print(f"make_dist: Downloading dist tarball from {download_url} ...")
+    request = urllib.request.Request(download_url, headers={"Authorization": "token " + api.token})
+    zipio = io.BytesIO()
+    try:
+        with urllib.request.urlopen(request) as response:
+            if os.isatty(sys.stdout.fileno()):
+                total_size = 0
+            else:
+                total_size = None
+            MB = 2**20
```

Comment:
i think you meant `1e6` :)

Reply:
That's a float. But I can give you 10**6 :grin: (downloading in MB or MiB blocks should make no practical difference)

Reply:
It's entirely about what we show the user